Initial fix node Hierarchy change when it is deactivating issue

ColinCollins opened the pull request:
Re: #2359

Changelog:
 1.add a new judgement in base-node.js setParent function 
2.add a new errorID number //Game 3000 part '3821'
 

<!-- Note: Makes sure these boxes are checked before submitting your PR - thank you!

- [ ] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
- [ ] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.

- To official teams:
  - [ ] Check that your javascript is following our [style guide](https://github.com/cocos-creator/fireball/blob/dev/.github/CONTRIBUTING.md) and end files with a newline
  - [ ] Document new code with comments in source code based on [API Docs](https://github.com/cocos-creator/fireball#api-docs)
  - [ ] Make sure any runtime log information in `cc.log` , `cc.error` or `new Error('')` has been moved into `DebugInfos.js` with an ID, and use `cc.logID(id)` or `new Error(cc._getErrorID(id))` instead.

-->

pandamicro commented on cocos2d/core/utils/base-node.js:
```diff
@@ -360,6 +360,12 @@ var BaseNode = cc.Class({
             }
         }
+
+        if (CC_DEBUG && !this._activeInHierarchy){
```
应该使用 _objFlags 来进行判断

pandamicro commented on cocos2d/core/utils/base-node.js:
```diff
@@ -360,6 +361,10 @@ var BaseNode = cc.Class({
                 return;
             }
         }
+        if (CC_DEBUG && (this._objFlags & Deactivating)) {
```
你可以想一下，在做 setParent 的时候，改变了 children 数组内容的是哪些对象？是 this 吗？

ColinCollins replied:
那我就还是在昨天那个循环的位置进行修改,循环内容增加对数组变化的处理

ColinCollins replied:
Deactivating判断下面的的数组变化的部分还是要保留,jare哥说要保证原先内容的正常运行,只是在需要报错的地方给个报错,那这样讲error部分的需要修改的部分就是从循环长度那边做了

pandamicro replied:
不是这个意思，setParent 是将 this 从 old parent 的 children 列表移除，再加入到 value（new parent）的 children 列表中，你这里应该判断的是 old 和 new parent 是否在 desactivating

jareguo commented on cocos2d/core/utils/base-node.js:
```diff
@@ -360,16 +361,26 @@ var BaseNode = cc.Class({
+        }
+        if (CC_DEBUG && (oldParent._objFlags & Deactivating)) {
+            cc.errorID(3821);
+        }
```
上面两个 if 合并成一个吧。既然情况这么复杂，请把单元测试加上。单元测试里 spy 一下 cc.errorID

pandamicro commented on cocos2d/core/utils/base-node.js:
```diff
@@ -360,16 +361,26 @@ var BaseNode = cc.Class({
+        if (CC_DEBUG && (this._objFlags & Deactivating)) {
+            cc.errorID(3821);
+        }
+        if (CC_DEBUG && (oldParent._objFlags & Deactivating)) {
```
我又想了一下，判断 this._objFlags 已经包含了对 oldParent 的判断了，因为 this._objFlags 一定是通过 deactivateNodeRecursively 传递过来的，所以不需要判断 oldParent，保留 this 和 value 的判断即可

ColinCollins replied:
这里的this,指的是最后一种单独调节当前node 的情况啊,如果是从oldParent传递过来的,会检测不到这种情况,就是单独的调用了SetParent 而没有运行deactivateNodeRecursively这个函数部分

pandamicro replied:
oldParent 传递过来的时候，this._objFlags & Deactivating 也一定是 true

jareguo replied:
来来来，改成测试驱动开发

jareguo replied:
搜一下 "disable component during onEnable"，加一个 "change parent during onDisable"

pandamicro commented on test/qunit/unit-es5/test-component-scheduler.js:
```diff
@@ -165,6 +165,85 @@ test('disable component during onEnable', function () {
+    nodes2.attachToScene();
+
+    testNode.active = false;
+    cc.errorID.once('Should not change owned Hierarchy during deactivating');
```
Should report error when changing target node's parent during its deactivation

pandamicro commented on test/qunit/unit-es5/test-component-scheduler.js:
```diff
@@ -165,6 +165,85 @@ test('disable component during onEnable', function () {
+    nodes3.attachToScene();
+
+    testNode.active = false; 
+    cc.errorID.once('Should not change the others\' Hierarchy to any deactivating node');
```
Should report error when append new child to a target node during its deactivation

pandamicro commented on test/qunit/unit-es5/test-component-scheduler.js:
```diff
@@ -165,6 +165,85 @@ test('disable component during onEnable', function () {
+    childComp = nodes1.childComps[0];
+   
+    childComp.onDisable = function() {
+        child.parent = preNode;
```
应该是在 nodes1 deactivating 的过程中，在 testNode 的 comp.onDisable 中让 child.parent = nodes1。

现在这样测试的情况其实和第一种本质是一样

pandamicro replied:
我建议你把第一个改成通过 node1.active = false 来触发，就够了

ColinCollins replied:
修正,第一第二测试的时候声明内容为:
1.deactivating 状态下节点不允许更改node层级
2.deactivating 状态下的节点不允许更改为其他node的父节点
这样的话,更改的第三种测试内容和第二种本质上就是一样了

pandamicro commented on test/qunit/unit-es5/test-component-scheduler.js:
```diff
@@ -165,6 +165,85 @@ test('disable component during onEnable', function () {
+    nodes1.attachToScene();
+
+    testNode.active = false;
+    cc.errorID.once('Should not change owned Hierarchy during deactivating');
```
Should report error when changing target node's parent during its deactivation

pandamicro commented on test/qunit/unit-es5/test-component-scheduler.js:
```diff
@@ -165,6 +165,85 @@ test('disable component during onEnable', function () {
+    nodes2.attachToScene();
+
+    testNode.active = false; 
+    cc.errorID.once('Should not change the others\' Hierarchy to any deactivating node');
```
Should report error when append new child to a target node during its deactivation

pandamicro commented on test/qunit/unit-es5/test-component-scheduler.js:
```diff
@@ -165,6 +165,85 @@ test('disable component during onEnable', function () {
+    nodes3.attachToScene();
+ 
+    testNode.active = false;  
+    cc.errorID.once('Should not use setSilingIndex to change\' Hierarchy during deactivating'); 
```
Should report error when setting node's sibling index during its parent's deactivation

ColinCollins committed:
forbid Hierarchy change test example 

add a  new test in test-component-scheduler.js
forbid Hierarchy change part and example

update data make my local file get reset
forbid Hierarchy change part test example

update a  new test content about setSiblingIndex
change the report of error test
